### PR TITLE
Support forwarding configurable headers through the gRPC-Web API.

### LINKED ISSUE #5999: Unauthorized response in GetPackageRepositorySummaries
```
Invalid GetPackageRepositorySummaries response from the plugin helm.packages: rpc error: code = Unauthenticated desc = Authorization required to get the AppRepository 'all' due to 'Unauthorized'
```

We're trying to upgrade to kubeapps 2.6.3 (latest release) and I've encountered a strange bug when trying to load the `Catalog`.

Seemingly due to an Unauthorized GRPC response, the dashboard redirects back to `/` and I end up right where I started (i.e. main dashboard page viewing applications).

1. Load kubeapps
2. Click `Catalog`
3. Redirected back to dashboard

image
image
image

Is there some new configuration now that we need to do to permit our users to view package repositories?

We have 2 `AppRepository` configured in this cluster (BTW what is the difference to the "PackageRepository" concept?):

```
(⎈ default:kubeapps)mcampbell-1 :: coreweave/k8s-services/dev ‹master*› » k neat get apprepository
apiVersion: v1
items:
- apiVersion: kubeapps.com/v1alpha1
  kind: AppRepository
  metadata:
    annotations:
      meta.helm.sh/release-name: kubeapps
      meta.helm.sh/release-namespace: kubeapps
    labels:
      app.kubernetes.io/instance: kubeapps
      app.kubernetes.io/managed-by: Helm
      app.kubernetes.io/name: kubeapps
      helm.sh/chart: kubeapps-12.2.1-dev0
    name: ant-repo
    namespace: kubeapps
  spec:
    syncJobPodTemplate:
      spec:
        securityContext:
          runAsUser: 1001
    type: helm
    url: http://charts.tenant-rizzo.ord1.ingress.coreweave.cloud/
- apiVersion: kubeapps.com/v1alpha1
  kind: AppRepository
  metadata:
    annotations:
      meta.helm.sh/release-name: kubeapps
      meta.helm.sh/release-namespace: kubeapps
    labels:
      app.kubernetes.io/instance: kubeapps
      app.kubernetes.io/managed-by: Helm
      app.kubernetes.io/name: kubeapps
      helm.sh/chart: kubeapps-12.2.1-dev0
    name: coreweave
    namespace: kubeapps
  spec:
    syncJobPodTemplate:
      spec:
        securityContext:
          runAsUser: 1001
    type: helm
    url: http://helm.corp.ingress.ord1.coreweave.com/
kind: List
metadata:
  resourceVersion: ""
  selfLink: ""
```

### PR DESCRIPTION
**Pull request by absoludity**
### Description of the change

Adds a plugin config option for the kubeapps API server so that headers in gRPC(-Web) requests are translated into headers on the request from the kubeapps API server to the k8s API server.

### Benefits

Enables functionality that used to work before we switched to the gRPC api server (as the dashboard used to talk directly to the k8s cluster, the headers were simply present).

### Applicable issues

- fixes #5999 

### Additional information

I've not tested this IRL yet, but feel free to try, otherwise I'll setup an IRL test-scenario tomorrow.

### CODE REVIEW

**Review by ppbaena — APPROVED** [review]

File: cmd/kubeapps-apis/core/plugins/v1alpha1/plugins_test.go [hunk 1]
```diff
@@ -431,4 +435,114 @@ func TestCreateConfigGetterWithParams(t *testing.T) {
 			}
 		})
 	}
+
+	// allow comparisons of unexported fields on teh headerAdder for this test.
```
Comment (mecampbellsoup) [inline cmd/kubeapps-apis/core/plugins/v1alpha1/plugins_test.go:439]: nit: the*

File: chart/kubeapps/values.yaml [hunk 2]
```diff
@@ -1561,6 +1561,8 @@ kubeappsapis:
             ## headerPattern: namespace:^([\w-]+):\w+$
             ##
             headerPattern: ""
+          ## @param kubeappsapis.pluginConfig.core.packages.v1alpha1.forwardedHeaders Optional list of headers that should be forwarded from the incoming gRPC-Web request to the k8s API server.
```
Comment (mecampbellsoup) [inline chart/kubeapps/values.yaml:1564]: i think this is at `kubeappsapis.pluginConfig.resources.packages.v1alpha1.forwardedHeaders`?

File: cmd/kubeapps-apis/plugins/resources/v1alpha1/common/plugin_test.go [hunk 3]
```diff
@@ -57,6 +57,23 @@ resources:
 			},
 			expectedError: "",
 		},
+		{
+			name: "parses forwarded headers config",
+			pluginYAMLConf: []byte(`
+resources:
+  packages:
+    v1alpha1:
```
Comment (mecampbellsoup) [inline cmd/kubeapps-apis/plugins/resources/v1alpha1/common/plugin_test.go:65]: missing `forwardedHeaders` key?

**Review by mecampbellsoup — COMMENTED** [review]
playing around w/ this locally and so far unable to get it working 😢 but figured i'd submit this review w/ some comments

**Comment by mecampbellsoup** [comment]
@absoludity so the request to `CheckNamespaceExists cluster="default" namespace="default"` is still failing even w/ the proxy auth headers included:

```
[kubeappsapis] I0222 14:01:44.208350       1 namespaces.go:27] "+resources CheckNamespaceExists" cluster="default" namespace="default"
[kubeappsapis] I0222 14:01:44.208426       1 plugins.go:361] "+[MATT] forwardedHeaders" forwardedHeaders=[X-Consumer-Username X-Consumer-Permissions]
[kubeappsapis] I0222 14:01:44.208458       1 plugins.go:414] "+[MATT] ctx" ctx="context.Background.WithValue(type *http.contextKey, val <not Stringer>).WithValue(type *http.contextKey, val 10.241.126.192:50051).WithCancel.WithCancel.WithCancel.WithValue(type metadata.mdIncomingKey, val <not Stringer>).WithValue(type peer.peerKey, val <not Stringer>).WithValue(type grpc.streamKey, val <not Stringer>)"
[kubeappsapis] I0222 14:01:44.208541       1 plugins.go:416] "+[MATT] metadata.FromIncomingContext" md=map[:authority:[kubeapps-internal-kubeappsapis:8080] accept:[*/*] accept-encoding:[gzip, deflate, br] accept-language:[en-US,en;q=0.9] authorization:[Bearer z4g1flvc38x9ylj50ybullmi0th4gd45] connection:[close] content-type:[application/grpc+proto] cookie:[csrftoken=ImlZdHRuRmNHWW0yUU5jMlQi.Z8aCXqWDNP5VbEyyjNqpWFFsJAo; sessionid=z4g1flvc38x9ylj50ybullmi0th4gd45] dnt:[1] origin:[https://apps.coreweave.test] referer:[https://apps.coreweave.test/] sec-ch-ua:["Chromium";v="110", "Not A(Brand";v="24", "Google Chrome";v="110"] sec-ch-ua-mobile:[?0] sec-ch-ua-platform:["Linux"] sec-fetch-dest:[empty] sec-fetch-mode:[cors] sec-fetch-site:[same-origin] sec-gpc:[1] user-agent:[Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/110.0.0.0 Safari/537.36] x-consumer-permissions:[w:org-24fdb0:users,r:org-24fdb0:users,r:cloud-app-dev:base,w:ns-tenant-dev-24fdb0-one:storage,w:ns-tenant-dev-24fdb0-one:virtualservers,r:ns-tenant-dev-24fdb0-one:full,r:ns-tenant-dev-24fdb0-one:base,w:ns-tenant-dev-24fdb0-one:objectstorage,r:ns-tenant-dev-24fdb0-one:storage,r:ns-tenant-dev-24fdb0-one:virtualservers,w:ns-tenant-dev-24fdb0-one:base,r:ns-tenant-dev-24fdb0-one:pods,r:ns-tenant-dev-24fdb0-one:objectstorage] x-consumer-username:[mcampbell+dev010@coreweave.com] x-forwarded-for:[10.10.0.6] x-forwarded-proto:[https] x-grpc-web:[1]] ok=true
[kubeappsapis] I0222 14:01:44.210079       1 server.go:62] Unauthenticated 1.741022ms /kubeappsapis.plugins.resources.v1alpha1.ResourcesService/CheckNamespaceExists
```
However as I'm thinking about it, this makes sense. We have a shared k8s "tenant cluster" which is shared by all of our customers, and we provision multi-tenancy via namespaces. So in the example above, user `mcampbell+dev010@coreweave.com` only has access (as you can sort of see via `x-consumer-permissions`) to namespace `tenant-dev-24fdb0-one`. That user cannot access the `default` namespace!

Do you know why the context `cluster="default" namespace="default"` is chosen for that request? Does it make sense for our use case?

File: chart/kubeapps/values.yaml (hunk 2 above)
Comment (absoludity) [reply chart/kubeapps/values.yaml:1564]: Woops, yes, I initially tried it in the core settings but thought it better here with the resources (which are required anyway).

**Comment by absoludity** [comment]
> @absoludity so the request to `CheckNamespaceExists cluster="default" namespace="default"` is still failing even w/ the proxy auth headers included:

Thanks for the logs - that's all I would have been able to test IRL anyway (that the headers get forwarded with the k8s request).

[...]

> 
> However as I'm thinking about it, this makes sense. We have a shared k8s "tenant cluster" which is shared by all of our customers, and we provision multi-tenancy via namespaces. So in the example above, user `mcampbell+dev010@coreweave.com` only has access (as you can sort of see via `x-consumer-permissions`) to namespace `tenant-dev-24fdb0-one`. That user cannot access the `default` namespace!
> 
> Do you know why the context `cluster="default" namespace="default"` is chosen for that request? Does it make sense for our use case?

Yes, the frontend dashboard is just checking if the user is authenticated to be able to check the default namespace, not if they're authorized to see the default namespace. That is, a 403 Unauthorized response would be fine here, as you can see in the code:

https://github.com/vmware-tanzu/kubeapps/blob/291d4d18a5aa30a010e3f4933cebdbe6880592a2/dashboard/src/shared/Auth.ts#L83-L98

but because an `Unauthenticated` response is being returned, the dashboard thinks there's an error. Is there a reason your proxy in front of your k8s API server is returning `Unauthenticated` rather than `Unauthorized` here? (I mean, the fact that the headers are present should mean that the request is authenticated right even if unauthorized, I think?)

**Comment by mecampbellsoup** [comment]
> (I mean, the fact that the headers are present should mean that the request is authenticated right even if unauthorized for this particular request, I think?)

This was a great question and provided for a promising lead 😄 

The issue seems to be that there is a lingering, leftover `Authorization` header in the request. Compare these 2, the first is how the request is being made as of the current code in this branch, the 2nd merely strips away the `Authorization: Bearer ...` header key-value.

```shell
root@cloud-app-6c97bb89fd-64r5d:/cloud-app# # With Authorization header
curl -X 'GET' \
  -H 'Accept: application/json, */*' \
  -H 'Authorization: Bearer odxu9rsy36z2tlmg29jf92lewqsxgni7' \
  -H 'User-Agent: kubeapps-apis/v0.0.0 (linux/amd64) kubernetes/$Format' \
  -H 'X-Consumer-Permissions: w:org-24fdb0:users,r:org-24fdb0:users,r:cloud-app-dev:base,w:ns-tenant-dev-24fdb0-one:storage,w:ns-tenant-dev-24fdb0-one:virtualservers,r:ns-tenant-dev-24fdb0-one:full,r:ns-tenant-dev-24fdb0-one:base,w:ns-tenant-dev-24fdb0-one:objectstorage,r:ns-tenant-dev-24fdb0-one:storage,r:ns-tenant-dev-24fdb0-one:virtualservers,w:ns-tenant-dev-24fdb0-one:base,r:ns-tenant-dev-24fdb0-one:pods,r:ns-tenant-dev-24fdb0-one:objectstorage' \
  -H 'X-Consumer-Username: mcampbell+dev010@coreweave.com' \
  'https://10.235.192.1:443/api/v1/namespaces/default' \
    --insecure
{"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"Unauthorized","reason":"Unauthorized","code":401}


root@cloud-app-6c97bb89fd-64r5d:/cloud-app# # Without Authorization header
curl -X 'GET' \
  -H 'Accept: application/json, */*' \
  -H 'User-Agent: kubeapps-apis/v0.0.0 (linux/amd64) kubernetes/$Format' \
  -H 'X-Consumer-Permissions: w:org-24fdb0:users,r:org-24fdb0:users,r:cloud-app-dev:base,w:ns-tenant-dev-24fdb0-one:storage,w:ns-tenant-dev-24fdb0-one:virtualservers,r:ns-tenant-dev-24fdb0-one:full,r:ns-tenant-dev-24fdb0-one:base,w:ns-tenant-dev-24fdb0-one:objectstorage,r:ns-tenant-dev-24fdb0-one:storage,r:ns-tenant-dev-24fdb0-one:virtualservers,w:ns-tenant-dev-24fdb0-one:base,r:ns-tenant-dev-24fdb0-one:pods,r:ns-tenant-dev-24fdb0-one:objectstorage' \
  -H 'X-Consumer-Username: mcampbell+dev010@coreweave.com' \
  'https://10.235.192.1:443/api/v1/namespaces/default' \
  --insecure
{"kind":"Status","apiVersion":"v1","metadata":{},"status":"Failure","message":"namespaces \"default\" is forbidden: User \"system:anonymous\" cannot get resource \"namespaces\" in API group \"\" in the namespace \"default\"","reason":"Forbidden","details":{"name":"default","kind":"namespaces"},"code":403}
```
And in the latter, there is our desired 403 response status!

**Comment by absoludity** [comment]
Closing this after discussion with Matt. If the kubeapps-apis server used the `apiServiceURL` instead of the default InCluster config, then their proxy should handle the rest correctly (adding those headers as required). The fact that kubeapps doesn't honor the `apiServiceURL` is a bug that I'll raise separately.

**Comment by absoludity** [comment]
Separate PR honoring the `apiServiceURL` for the cluster on which Kubeapps is installed: https://github.com/vmware-tanzu/kubeapps/pull/6016

**Pull request closed without merging**